server/src/main.rs

```diff
@@ -0,0 +1,201 @@
+
+        self.client
+            .log_message(MessageType::INFO, "Slice language server initialized!")
+            .await;
+    }
+
+    async fn shutdown(&self) -> tower_lsp::jsonrpc::Result<()> {
+        Ok(())
+    }
+
+    async fn goto_definition(
+        &self,
+        params: GotoDefinitionParams,
+    ) -> tower_lsp::jsonrpc::Result<Option<GotoDefinitionResponse>> {
+        let uri = params.text_document_position_params.text_document.uri;
+        let position = params.text_document_position_params.position;
+        let state = self.compilation_state.lock().await;
+
+        if let Some(location) = get_definition_span(&state, uri, position) {
+            let start = Position {
+                line: (location.start.row - 1) as u32,
+                character: (location.start.col - 1) as u32,
+            };
+
+            let end = Position {
+                line: (location.end.row - 1) as u32,
+                character: (location.end.col - 1) as u32,
+            };
+
+            Ok(Some(GotoDefinitionResponse::Scalar(Location {
+                uri: Url::from_file_path(location.file).unwrap(),
+                range: Range::new(start, end),
+            })))
+        } else {
+            Ok(None)
+        }
+    }
+
+    async fn hover(&self, params: HoverParams) -> tower_lsp::jsonrpc::Result<Option<Hover>> {
+        let uri = params.text_document_position_params.text_document.uri;
+        let position = params.text_document_position_params.position;
+        let state = self.compilation_state.lock().await;
+        if let Some(info) = get_hover_info(&state, uri, position) {
+            Ok(Some(Hover {
+                contents: HoverContents::Scalar(MarkedString::String(info)),
+                range: None,
+            }))
+        } else {
+            Ok(None)
+        }
+    }
+
+    async fn did_open(&self, params: DidOpenTextDocumentParams) {
+        self.client
+            .log_message(MessageType::INFO, "file opened!")
+            .await;
+        self.handle_file_change(params.text_document.uri).await;
+    }
+
+    async fn did_save(&self, params: DidSaveTextDocumentParams) {
+        self.client
+            .log_message(MessageType::INFO, "file saved!")
+            .await;
+        self.handle_file_change(params.text_document.uri).await;
+    }
+}
+
+impl Backend {
+    async fn handle_file_change(&self, uri: Url) {
+        let (updated_state, options) = self.compile_slice_files().await;
+        let mut compilation_state_guard = self.compilation_state.lock().await;
+        *compilation_state_guard = updated_state;
+
+        let mut compilation_options_guard = self.compilation_options.lock().await;
+        *compilation_options_guard = options;
+
+        let diagnostics = self
+            .get_diagnostics_for_uri(
+                &uri,
+                &mut compilation_state_guard,
+                &compilation_options_guard,
+            )
+            .await;
```

Comment:
Does this leave `compilation_state_guard` in a usable (stable) state once the mutex is unlocked?